### PR TITLE
RSE-1588: Add Pagination to Manage Workflow and Dashboard Page + New Apis

### PR DESCRIPTION
**Pull request by deb1990**
## Overview
As part of this PR, pagination has been added to the Manage Award and Dashboard page, to only show 25 records at a time.
This is related to https://github.com/compucorp/uk.co.compucorp.civicase/pull/692.

## Before/After
Screenshots can be found at https://github.com/compucorp/uk.co.compucorp.civicase/pull/692.

## Technical Details
**New Apis - Award.get and Award.getCount **
Previously, to filter Case Types by award specific details like manager or start_date etc, we had to call 3 APIs
1. AwardManager
2. AwardDetails
3. CaseType

and it was done from the FE, but it's not an efficient way. So created `Award.get` api, which accepts 3 params
1. managed_by - to filter using AwardManager API
2. award_detail_params - to filter using AwardDetails API
3. case_type_params - to filter using CaseType API

It processes all the filters in a single api call, and returns related case types. And `Award.getCount` just returns the number of records, useful for pagination.

**New Event Listener for Case.GetDetails**
Previously, when fetching Cases(using Case.Getdetails) or Activities(using `case_filter` which internally uses Case.Getdetails api), there was no option to include Award Specific filters(AwardManager and AwardDetail).
So now a new listener is added, which internally applies the `Award.get` api to filter by Awards too.
It extracts the following params
1. 'case_type_id.managed_by' -> to filter using AwardManager API
2. 'case_type_id.award_detail_params' -> to filter using AwardDetails API
3. Other params starting with 'case_type_id.'' -> to filter using CaseType API

**applicant-management-workflow.service.js**
1. `getActivityFilters` returns the initial filters to load the dashboard.
2. `getWorkflowsList` has been divided into 2 functions, one for getting normal workflows(has same name as previous), and another to get formatted results for Manage Workflows page(new name is `getFormattedWorkflowsList `).

**more-filters-dashboard-action-button.controller.js**
1. Instead of calling api to fetch CaseType IDs matching awards filters, it now just assigns the award filter params to the `$scope.activityFilters.case_filter` object. Then the filtering is handled globally CiviCase code base using newly created APIs.

### CODE REVIEW

**Comment by jamienovick** [comment]
@deb1990 @tunbola did we make a decision on how to have a global solution to this limit issue? Did we manage to get a ticket into the backlog for that? Thanks cc @james-compucorp @sohaib-compucorp

**Comment by tunbola** [comment]
@jamienovick , Yes we did agree to create a new extension to fix this issue (Based on the email thread). We will make the this extension a dependency for civicase and also add it in to compuclient. Other extensions too can depend on this new extension.
I'll discuss with @james-compucorp and add in a ticket for this so it can be prioritised accordingly.

**Comment by deb1990** [comment]
Yes Jamie, I was working on other tickets, so could not focus on this. The decision was to create a new extention, so once I am back to this ticket, we discuss how exactly to achieve this. 
Also if we want to include this in compuclient, there are some issues, like all other extensions needs to be updated, because otherwise all records will be shown, where no limit is mentioned by default, which can create unneeded results.
So we will discuss the complete plan, and proceed accordingly.

**Comment by deb1990** [comment]
@jamienovick @davialexandre @omarabuhussein 

I just had a discussion with @tunbola , and I have a concern with this approach.

If we add this Listener(Implicit limit), then what would happen to the Core Code, and in any 3rd party extensions, which expects 25 records by default(and limit 25 is not mentioned already). In all those places, all records will be shown, which will be an unexpected result.

According to @tunbola we did not face any issues in CiviHR, but compuclient is a much bigger system with many 3rd party extentions, so something breaking might be a possibility. So I wanted to flag this, before we start creating the new extention.

**Review by tunbola — CHANGES_REQUESTED** [review]

File: ang/civiawards-base/services/applicant-management-workflow.service.js [hunk 1]
```diff
@@ -103,15 +148,15 @@
...
 
     /**
-     * @param {object[]} workflows list of workflows
-     * @returns {object[]} formatted list of workflows
+     * @param {object[]} results results
+     * @returns {object[]} formatted results
      */
-    function getFormattedAwardDetailsData (workflows) {
-      var managerContacts = getAwardManagersForWorkflows(workflows);
+    function getFormattedAwardDetailsData (results) {
```
Comment (reneolivo) [inline ang/civiawards-base/services/applicant-management-workflow.service.js:154]: there are a couple of things here:

* `getFormattedWorkflowsList` and `getWorkflowsList` should return an object instead of an array. `{get: {}, count: {}}` maybe? The reason is that it's difficult to know which index belongs to which result.
* `getFormattedAwardDetailsData` should only accept workflows as it was doing before, it should not know that the results is an array object containing workflows and counts. Since it won't do anything about the counts it should only accept workflows so we avoid confusing the purpose of the function.

I think `getFormattedWorkflowsList` and `getWorkflowsList` need to be refactored in a way similar to this:

```js
function getFormattedWorkflowsList (...) {
  // ...
  var apiCalls = {
    get: ['Award', 'get', paramsWithLimit],
    count: ['Award', 'getcount', params]
  };

  return civicaseCrmApi(apiCalls).then(function (data) {
    return {
      get: getFormattedAwardDetailsData(data.get.values),
      count: data.count
    };
  });
}
```

File: ang/civiawards/dashboard/directives/more-filters-dashboard-action-button.controller.js [hunk 2]
```diff
@@ -86,31 +77,32 @@
...
+
+      if (model.selectedFilters.managed_by !== 'all_awards') {
+        $scope.activityFilters.case_filter['case_type_id.managed_by'] = model.selectedFilters.managed_by;
+      } else {
+        delete $scope.activityFilters.case_filter['case_type_id.managed_by'];
+      }
+
+      $scope.activityFilters.case_filter['case_type_id.award_detail_params'] = getAwardApiFilters();
+
+      $rootScope.$broadcast('civicase::dashboard-filters::updated');
```
Comment (reneolivo) [inline ang/civiawards/dashboard/directives/more-filters-dashboard-action-button.controller.js:105]: @deb1990 isn't this updating the filters and triggering multiple requests?

I thought we were going to include this into a service that can be requested by the dashboard directive before it executed the request to fetch the workflows.

**Review by reneolivo — CHANGES_REQUESTED** [review]

File: api/v3/Award.php [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+ * @return array
+ *   API result descriptor
+ */
+function civicrm_api3_award_get(array $params) {
+  $awardsFilteredByManager = [];
+  if ($params['managed_by']) {
+    $awardsFilteredByManager = civicrm_api3('AwardManager', 'get', [
+      'sequential' => 1,
+      'contact_id' => $params['managed_by'],
+      'options' => ['limit' => 0],
```
[the remark this answers is not in the document]
Comment (deb1990) [reply api/v3/Award.php]: As discussed, this is fine for now, I will add a ticket to refactor the backend logic later.

File: api/v3/Award.php [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+ *   API parameters.
+ *
+ * @return array
+ *   API result descriptor
+ */
+function civicrm_api3_award_getcount(array $params) {
+  $params['case_type_params']['options']['limit'] = 0;
+  $params['case_type_params']['options']['offset'] = 0;
+
+  return count(civicrm_api3_award_get($params)['values']);
```
[the remark this answers is not in the document]
Comment (deb1990) [reply api/v3/Award.php:156]: As discussed, this is fine for now, I will add a ticket to refactor the backend logic later.

File: CRM/CiviAwards/Event/Listener/AwardCaseFilter.php [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+      if ($param === 'case_type_id.managed_by' || $param === 'case_type_id.award_detail_params') {
+        unset($apiRequest['params'][$param]);
+      }
+      elseif (strpos($param, 'case_type_id.') === 0) {
+        unset($apiRequest['params'][$param]);
+        $caseTypeParams[substr($param, strlen('case_type_id.'))] = $value;
+      }
+    }
+
+    $filteredAwardTypes = civicrm_api3('Award', 'get', [
```
[the remark this answers is not in the document]
Comment (deb1990) [reply CRM/CiviAwards/Event/Listener/AwardCaseFilter.php]: As discussed, this is fine for now, I will add a ticket to refactor the backend logic later.

File: ang/civiawards-base/services/applicant-management-workflow.service.js [hunk 6]
```diff
@@ -32,56 +47,86 @@
...
+
+      var apiCalls = [
+        ['Award', 'get', paramsWithLimit],
+        ['Award', 'getcount', params]
+      ];
+
+      return civicaseCrmApi(apiCalls).then(function (data) {
+        return getFormattedAwardDetailsData(data);
+      });
+    }
```
[the remark this answers is not in the document]
Comment (deb1990) [reply ang/civiawards-base/services/applicant-management-workflow.service.js:104]: One is for Case overview page, and the other is for Manage Workflows. They need to have some different pre and post process. But I agree that the naming is confusing. So now I named them `getWorkflowsListForCaseOverview` and `getWorkflowsListForManageWorkflow`. Also created more function to remove duplicate logic.

File: ang/civiawards/dashboard/directives/more-filters-dashboard-action-button.controller.js (hunk 2 above)
Comment (deb1990) [reply ang/civiawards/dashboard/directives/more-filters-dashboard-action-button.controller.js:105]: No its not, because previously the initial logic to Apply My Awards was being applied using this Controller. But now it does not do anything initially.
Now https://github.com/compucorp/uk.co.compucorp.civiawards/pull/177/files#diff-64f7e3308246b806bc6b1307306016ee6c0acd0c249fd17763e9cd63f3228391R35 does that. That way we dont have to make multiple calls.

> I thought we were going to include this into a service that can be requested by the dashboard directive before it executed the request to fetch the workflows.

This was tough to implement, as currently we a controller for every dashboard action button, and not a service. So there is no way to apply functions before init without refactoring the code a lot.

File: api/v3/Award.php [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+/**
+ * Award.get API.
+ *
+ * @param array $params
+ *   API parameters.
+ *
+ * @return array
+ *   API result descriptor
+ */
+function civicrm_api3_award_get(array $params) {
```
[the remark this answers is not in the document]
Comment (deb1990) [reply api/v3/Award.php:43]: It will take me more time to understand and write tests, and I dont want to spend more time in this ticket. But definitely, I will start adding BE tests in the future.

Will add a ticket though 👍

File: CRM/CiviAwards/Service/AwardCaseFilterPreProcess.php [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+ */
+class CRM_CiviAwards_Service_AwardCaseFilterPreProcess {
+
+  /**
+   * Function that handles pre-processing for a case details relating to awards.
+   *
+   * @param array $requestParams
+   *   API request parameters.
+   */
+  public function onCreate(array &$requestParams) {
```
Comment (tunbola) [inline CRM/CiviAwards/Service/AwardCaseFilterPreProcess.php:14]: This name actually depicts the event on which this is fired.  The event on which this is called is not related to the creation of an API. Maybe we can rename to `onCaseDetailsGet` or better still `alterRequestParams` to depict what the function is doing.

**Review by tunbola — CHANGES_REQUESTED** [review]

**Review by reneolivo — APPROVED** [review]

**Review by tunbola — DISMISSED** [review]

**Review by tunbola — APPROVED** [review]
